lib/include/openamp/virtio.h

```diff
@@ -161,6 +161,11 @@ void virtio_describe(struct virtio_device *dev, const char *msg,
  */
 
 struct virtio_dispatch {
+	int (*create_virtqueues)(struct virtio_device *vdev,
+				 unsigned int flags,
+				 unsigned int nvqs, const char *names[],
+				 vq_callback callbacks[]);
+	void (*delete_virtqueues)(struct virtio_device *vdev);
```

Comment:
> But I see the virtqueues create and delete in linux are in transport layer. See https://github.com/torvalds/linux/blob/master/drivers/virtio/virtio_mmio.c#L490 and https://github.com/torvalds/linux/blob/master/drivers/virtio/virtio_mmio.c#L354

Ok i start to understand your need... 

The equivalent of the `virtio_dispatch` is the [`virtio_config_ops`](https://elixir.bootlin.com/linux/latest/source/include/linux/virtio_config.h#L106)

so you need an equivalent of `find_vqs` and `del_vqs`

Now I would like to understand why the backends can not call `virtio_create_virtqueues` instead of duplicating the code ?

Reply:
I'm busy with other projects recently and very sorry for the late reply.

> The equivalent of the virtio_dispatch is the [virtio_config_ops]
> so you need an equivalent of find_vqs and del_vqs

Yes, I reference the Linux design.

> Now I would like to understand why the backends can not call virtio_create_virtqueues instead of duplicating the code ?

Because different transport layer has different implementation, e.g. mmio transport layer need set the virtqueue address but the remoteproc do not need.

Reply:
> I'm busy with other projects recently and very sorry for the late reply.

No worries, It is thesame for everyone ;-)

> > Now I would like to understand why the backends can not call virtio_create_virtqueues instead of duplicating the code ?
> 
> Because different transport layer has different implementation, e.g. mmio transport layer need set the virtqueue address but the remoteproc do not need.

If you have some code to share, I'm interested in. That would help me to understand your point

Reply:
@arnopo I think you get the sample code from here:
https://github.com/OpenAMP/open-amp/pull/494/commits/ae651f3c9970fe12dd00c16c25eff5e5b64596e8#r1227687251
@CV-Bowen just extract the core change from @danmilea and make it generalize to support both remoteproc and mmio(or pci) transport backend.